File: lib/providers/digikey.ts
import { ProviderResult } from "@/types/bom";
import { PartProvider } from "./types";

const DIGIKEY_TOKEN_URL = "https://api.digikey.com/v1/oauth2/token";
const DIGIKEY_SEARCH_URL = "https://api.digikey.com/products/v4/search/keyword";

// Token cache for OAuth
let cachedToken: { token: string; expiresAt: number } | null = null;

interface DigiKeySearchResponse {
  Products: DigiKeyProduct[];
  ProductsCount: number;
}

interface DigiKeyProduct {
  DigiKeyPartNumber: string;
  ManufacturerPartNumber: string;
  Manufacturer: { Name: string };
  ProductDescription: string;
  DetailedDescription: string;
  ProductUrl: string;
  QuantityAvailable: number;
  UnitPrice: number;
  MinimumOrderQuantity: number;
  StandardPricing: Array<{
    BreakQuantity: number;
    UnitPrice: number;
  }>;
}

export class DigiKeyProvider implements PartProvider {
  name = "digikey";
  private clientId: string;
  private clientSecret: string;

  constructor() {
    this.clientId = process.env.DIGIKEY_CLIENT_ID || "";
    this.clientSecret = process.env.DIGIKEY_CLIENT_SECRET || "";
  }

  private async getAccessToken(): Promise<string> {
    // Check if we have a valid cached token (with 60s buffer)
    if (cachedToken && cachedToken.expiresAt > Date.now() + 60000) {
      return cachedToken.token;
    }

    if (!this.clientId || !this.clientSecret) {
      throw new Error("DigiKey OAuth credentials not configured");
    }

    const response = await fetch(DIGIKEY_TOKEN_URL, {
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
      },
      body: new URLSearchParams({
        grant_type: "client_credentials",
        client_id: this.clientId,
        client_secret: this.clientSecret,
      }),
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`DigiKey OAuth token request failed: ${response.status} - ${errorText}`);
    }

    const data = await response.json();

    // Cache the token
    cachedToken = {
      token: data.access_token,
      expiresAt: Date.now() + (data.expires_in * 1000),
    };

    console.log("Obtained new DigiKey OAuth token");
    return cachedToken.token;
  }

  async search(query: string): Promise<ProviderResult[]> {
    return this.executeSearch(query);
  }

  async searchByMPN(mpn: string, manufacturer?: string): Promise<ProviderResult[]> {
    return this.executeSearch(mpn);
  }

  private async executeSearch(query: string): Promise<ProviderResult[]> {
    if (!this.clientId || !this.clientSecret) {
      throw new Error("DigiKey OAuth credentials not configured. Set DIGIKEY_CLIENT_ID and DIGIKEY_CLIENT_SECRET.");
    }

    const token = await this.getAccessToken();

    const response = await fetch(DIGIKEY_SEARCH_URL, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
        "X-DIGIKEY-Client-Id": this.clientId,
      },
      body: JSON.stringify({
        Keywords: query,
        RecordCount: 10,
        RecordStartPosition: 0,
        ExcludeMarketPlaceProducts: true,
      }),
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`DigiKey API error: ${response.status} - ${errorText}`);
    }

    const data: DigiKeySearchResponse = await response.json();
    return this.transformResults(data);
  }

  private transformResults(data: DigiKeySearchResponse): ProviderResult[] {
    const results: ProviderResult[] = [];

    if (!data.Products) {
      return results;
    }

    for (const product of data.Products) {
      // Get the best price (qty 1 or lowest tier)
      let price = product.UnitPrice;
      if (product.StandardPricing && product.StandardPricing.length > 0) {
        price = product.StandardPricing[0].UnitPrice;
      }

      if (!price || price <= 0) continue;

      results.push({
        mpn: product.ManufacturerPartNumber,
        manufacturer: product.Manufacturer?.Name || "Unknown",
        description: product.ProductDescription || product.DetailedDescription || "",
        price,
        currency: "USD",
        stock: product.QuantityAvailable || 0,
        minQuantity: product.MinimumOrderQuantity || 1,
        provider: this.name,
        distributor: "DigiKey",
        url: product.ProductUrl || `https://www.digikey.com/products/en?keywords=${encodeURIComponent(product.ManufacturerPartNumber)}`,
      });
    }

    return results.sort((a, b) => a.price - b.price);
  }
}

File: python-agent/src/bom_agent_service/providers/digikey.py
"""DigiKey part provider with OAuth 2.0 authentication."""

import logging
import os
import time
from typing import Optional

import httpx

from .types import PartProvider, ProviderResult

logger = logging.getLogger(__name__)

DIGIKEY_TOKEN_URL = "https://api.digikey.com/v1/oauth2/token"
DIGIKEY_SEARCH_URL = "https://api.digikey.com/products/v4/search/keyword"


class TokenCache:
    """In-memory OAuth token cache."""

    def __init__(self):
        self.token: Optional[str] = None
        self.expires_at: float = 0


_token_cache = TokenCache()


class DigiKeyProvider(PartProvider):
    """DigiKey part provider using OAuth 2.0."""

    name = "digikey"

    def __init__(self):
        self.client_id = os.getenv("DIGIKEY_CLIENT_ID", "")
        self.client_secret = os.getenv("DIGIKEY_CLIENT_SECRET", "")

    def is_configured(self) -> bool:
        """Check if DigiKey credentials are configured."""
        return bool(self.client_id and self.client_secret)

    async def _get_access_token(self) -> str:
        """Get OAuth access token with caching."""
        # Check if we have a valid cached token (with 60s buffer)
        if _token_cache.token and _token_cache.expires_at > time.time() + 60:
            return _token_cache.token

        if not self.client_id or not self.client_secret:
            raise ValueError("DigiKey OAuth credentials not configured")

        async with httpx.AsyncClient() as client:
            response = await client.post(
                DIGIKEY_TOKEN_URL,
                data={
                    "grant_type": "client_credentials",
                    "client_id": self.client_id,
                    "client_secret": self.client_secret,
                },
                headers={"Content-Type": "application/x-www-form-urlencoded"},
            )
            response.raise_for_status()
            data = response.json()

            _token_cache.token = data["access_token"]
            _token_cache.expires_at = time.time() + data["expires_in"]

            logger.info("Obtained new DigiKey OAuth token")
            return _token_cache.token

    async def search(self, query: str) -> list[ProviderResult]:
        """Search for parts by query."""
        return await self._execute_search(query)

    async def search_by_mpn(
        self, mpn: str, manufacturer: Optional[str] = None
    ) -> list[ProviderResult]:
        """Search for parts by MPN."""
        return await self._execute_search(mpn)

    async def _execute_search(self, query: str) -> list[ProviderResult]:
        """Execute search against DigiKey API."""
        if not self.is_configured():
            raise ValueError(
                "DigiKey credentials not configured. "
                "Set DIGIKEY_CLIENT_ID and DIGIKEY_CLIENT_SECRET."
            )

        token = await self._get_access_token()

        async with httpx.AsyncClient() as client:
            response = await client.post(
                DIGIKEY_SEARCH_URL,
                headers={
                    "Authorization": f"Bearer {token}",
                    "X-DIGIKEY-Client-Id": self.client_id,
                    "Content-Type": "application/json",
                },
                json={
                    "Keywords": query,
                    "RecordCount": 10,
                    "RecordStartPosition": 0,
                    "ExcludeMarketPlaceProducts": True,
                },
                timeout=30.0,
            )
            response.raise_for_status()
            return self._transform_results(response.json())

    def _transform_results(self, data: dict) -> list[ProviderResult]:
        """Transform DigiKey response to ProviderResult list."""
        results: list[ProviderResult] = []

        products = data.get("Products", [])
        if not products:
            return results

        for product in products:
            # Get the best price (qty 1 or lowest tier)
            price = product.get("UnitPrice", 0)
            standard_pricing = product.get("StandardPricing", [])
            if standard_pricing:
                price = standard_pricing[0].get("UnitPrice", price)

            if not price or price <= 0:
                continue

            manufacturer_data = product.get("Manufacturer", {})
            mpn = product.get("ManufacturerPartNumber", "")

            results.append(
                ProviderResult(
                    mpn=mpn,
                    manufacturer=manufacturer_data.get("Name", "Unknown"),
                    description=product.get("ProductDescription", "")
                    or product.get("DetailedDescription", ""),
                    price=price,
                    currency="USD",
                    stock=product.get("QuantityAvailable", 0),
                    min_quantity=product.get("MinimumOrderQuantity", 1),
                    provider=self.name,
                    distributor="DigiKey",
                    url=product.get("ProductUrl", "")
                    or f"https://www.digikey.com/products/en?keywords={mpn}",
                )
            )

        return sorted(results, key=lambda r: r.price)
